### ISSUE TITLE
Enable static graph restore in build

### ISSUE THREAD

**Issue by danmoseley** [dotnet/aspnetcore#47808]
### Is there an existing issue for this?

- [X] I have searched the existing issues

### Describe the bug

A recent PR in another repo similar to the below enabled static graph restore. Apparently right now that only speeds up the restore phase. Any reason @rainersigwald we shouldn't switch it on here?

```
+++ b/Directory.Build.props

+    <!--
+      Opt Arcade-driven (restore.cmd, official build, and CI build) restores into using NuGet's static-graph restore
+      (https://learn.microsoft.com/nuget/reference/msbuild-targets#restoring-with-msbuild-static-graph-evaluation),
+      which uses MSBuild's static-graph-construction features to speed up the restore operation.
+     -->
+    <RestoreUseStaticGraphEvaluation>true</RestoreUseStaticGraphEvaluation>
+
     <!-- These are explicitly set in src/Analyzers and src/Generators -->
     <GenerateNeutralResourcesLanguageAttribute>false</GenerateNeutralResourcesLanguageAttribute>

+++ b/eng/Build.props
@@ -0,0 +1,5 @@
+<Project>
+  <PropertyGroup>
+    <RestoreUseStaticGraphEvaluation Condition="'$(DotNetBuildFromSource)' != 'true'">true</RestoreUseStaticGraphEvaluation>
+  </PropertyGroup>
+</Project>
```

**Comment by rainersigwald** [comment]
I'm not aware of any reason not to. @jeffkl?

**Comment by jeffkl** [comment]
As long as you're not using VS 17.5 where its still broken, then yes.  Be aware that hosted builds usually pull packages making restores slower in general so static graph-based restore won't make a huge difference.  But for local no-op restores it is way faster.

**Comment by wtgodbe** [comment]
@mmitche was this something you already experimented with in this repo? I think I remember you opening a draft PR for it within the past few weeks.

**Comment by mmitche** [comment]
I tried. It broke some things, didn't look much further. The build times didn't seem too much faster: https://github.com/dotnet/aspnetcore/pull/47813

**Comment by mkArtakMSFT** [comment]
Given that there were no major perf benefits, I think we should just not do it then. @mmitche, @danmoseley  is there a counterargument against this plan?

**Comment by danmoseley** [comment]
Nope, I have no opinion, just heard it might be faster. Although, @jeffkl comment above suggests it's a win mainly on dev machines -- @mmitche did you time locally or just in the PR validation?

**Comment by jeffkl** [comment]
Yes static graph restore just makes it much faster for NuGet to parse all of the projects but doesn't make downloading packages any faster.  So hosted builds on clean machines won't see a huge gain but locally when you're doing implicit restores before builds on large project trees, there are big gains.  When testing against 700+ projects, no-op restore dropped from 30 seconds to 9.  So the more projects you have, the more performance gains you'll see.

**Comment by rainersigwald** [comment]
This came up again and I poked at it gently. What I observed is that the Arcade `Build.proj` calls `Restore` on _each individual project_ because aspnetcore uses `ProjectToBuild` instead of a central solution . . . and that means that each individual project got its own static graph restore process running, which didn't help anything. To move forward I think we'd need to change Arcade to get that compatible with NuGet/MSBuild static graph construction.
